# 微分

改めてきちんと定義しますが、関数を$f(x)$とおいたときに、微分は次のように定義されます。このときの $f'(x)$ を**導関数**といいます。

$$f'(x) = \lim_{h \rightarrow 0} \frac{ f(x+h) - f(x) }{ h }$$

仮に $f(x)$ を $f(x) = x^2$ としたときに、コードでは次のようになります。

In [52]:
using Printf

# hを限りなく小さい値に設定
h = 0.000000000001


# f(x) を定義
function f(x)
    return x^2
end

# 微分計算
function diff_f(x)
    diff = f(x+h) - f(x)
    return diff / h
end


diff_f (generic function with 1 method)

これを仮に $x=2$ で実行してみましょう。

In [53]:
println(diff_f(2))

4.000355602329364


この$4.000355602329364$を$f(x)$の$x=2$における**微分係数**といいます。

これで関数$f(x)=x^2$　に対して導関数 $f'(x)$を求め、$x=2$における微分係数を計算しました。

## 微分の解説

### 平均変化率の定義

微分の解説に入る前に、中学校で習う**平均変化率**を確認します。平均変化率は、関数のある2点における変化の割合です。

#### 定義: 平均変化率の定義

関数$f(x)$に対して2点$A (a,f(a)), B (b,f(b))$の平均変化率は

$$ 
\begin{aligned}
\frac{yの変化量}{xの変化量} &= \frac{f(b) - f(a)}{ b - a}
\end{aligned}
$$
である。


<img src="./images/01/平均変化率.png" alt="Drawing" style="width: 500px;"/>


平均変化率は2点を結んだ直線の**傾き**です。今回の場合は、$(-1, 1)$と $(2, 4)$の平均変化率なので

$$
\frac{4 - 1}{2 - (-1)} = 1
$$

です。

平均変化率から2点が通る直線を求める式は、

$$y - f(a) = \frac{f(b) - f(a)}{ b - a }(x - a) $$

もしくは

$$y - f(b) = \frac{f(b) - f(a)}{ b - a }(x - b) $$

で求められます。


ここで平均変化率を求めるコードを紹介します。


In [54]:
# 平均変化率
function change_ratio(a, b)
    diff_y = f(b) - f(a)
    diff_x = b - a
    return diff_y / diff_x
end

println("微分係数: ", change_ratio(-1, 2))

微分係数: 1.0


### 平均変化率のパターン 

平均変化率は3つに大別できます。

<img src="./images/01/平均変化率のパターン.png" alt="Drawing" style="width: 500px;"/>

平均変化率が正の値のときは増加傾向、負のときは減少傾向、0のときは一定の傾向です。

###　平均変化率と微分

平均変化率と微分は大いに関係があります。平均変化率は**2点における変化率**でしたが、微分は**1点における変化率**です。

ここで、先ほどの平均変化率を求める式の形を変えてみましょう。

<img src="./images/01/bの表現.png" alt="Drawing" style="width: 500px;"/>


点$A$の$x$座標と点$B$の$x$座標の差を$h$とすると、$h = b - a$です。式変形するは$b$は$b=a+h$と表現できます。
この表現を平均変化率の式に適用してみましょう。

$$
\begin{aligned}
平均変化率 &= \frac{f(a+h) - f(a)}{ (a+h) - a }\\
         &= \frac{f(a+h) - f(a)}{h}   
\end{aligned}
$$

最初に紹介した微分の式にだいぶ近づいてきました。$h$は$a$と$b$の差分でしたので、「$h$を限りなく$0$に近づける」ことは「$b$を限りなく$a$に近づける」、すなわち「関数$f(x)$の点$x=a$における変化率」を求めることに等しいことになります。

$f(x)=x^2$の$x=-1$における微分係数を求めてみましょう。上の平均変化率の式に$f(x)=x^2$と$a=-1$を代入します。

$$
\begin{aligned}
\frac{f(-1+h) - f(-1)}{h} &= \frac{ (-1+h)^2 - (-1)^2 }{h}\\
    &= \frac{(1 - 2h + h^2) - 1}{h}\\
    &= \frac{-2h + h^2}{h}\\
    &=  -2 + h
\end{aligned}
$$


この　$h$を限りなく$0$に近づけると、


$$\lim_{h \rightarrow 0} \frac{f(-1+h) - f(-1)}{h} = -2$$

と、微分係数が$-2$になることがわかります。


実際にグラフで確認してみましょう。



グラフを見ると  で接していることがわかります。微分係数がであることとグラフでも確認できる通り、のときはの地点では減少傾向となっています。

今度はJuliaで、少しずつ$b$を$a$に近づけていって平均変化率が-2に近づいていくのを確認してみましょう。


In [55]:
a = -1.0
b = 2.0

# aとbの分割数。増やせば増やすほど最終的にaに近づく。
MAX=100

for n=range(1,MAX, step=1)
    h = (b - a) / n
    @printf("a=%0.6f,b=%0.6fのときの変化率は%0.6f\n", a, a+h, change_ratio(a, a+h))
end

a=-1.000000,b=2.000000のときの変化率は1.000000
a=-1.000000,b=0.500000のときの変化率は-0.500000
a=-1.000000,b=0.000000のときの変化率は-1.000000
a=-1.000000,b=-0.250000のときの変化率は-1.250000
a=-1.000000,b=-0.400000のときの変化率は-1.400000
a=-1.000000,b=-0.500000のときの変化率は-1.500000
a=-1.000000,b=-0.571429のときの変化率は-1.571429
a=-1.000000,b=-0.625000のときの変化率は-1.625000
a=-1.000000,b=-0.666667のときの変化率は-1.666667
a=-1.000000,b=-0.700000のときの変化率は-1.700000
a=-1.000000,b=-0.727273のときの変化率は-1.727273
a=-1.000000,b=-0.750000のときの変化率は-1.750000
a=-1.000000,b=-0.769231のときの変化率は-1.769231
a=-1.000000,b=-0.785714のときの変化率は-1.785714
a=-1.000000,b=-0.800000のときの変化率は-1.800000
a=-1.000000,b=-0.812500のときの変化率は-1.812500
a=-1.000000,b=-0.823529のときの変化率は-1.823529
a=-1.000000,b=-0.833333のときの変化率は-1.833333
a=-1.000000,b=-0.842105のときの変化率は-1.842105
a=-1.000000,b=-0.850000のときの変化率は-1.850000
a=-1.000000,b=-0.857143のときの変化率は-1.857143
a=-1.000000,b=-0.863636のときの変化率は-1.863636
a=-1.000000,b=-0.869565のときの変化率は-1.869565
a=-1.000000,b=-0.875000のときの変化率は-1.875000
a=-1.000000,b=-0.880

100回の思考だとまだ遠いですが、回数を増やせば増やすほど$-2$に近づいていきます。

ここで改めて、正確な導関数の定義をしましょう。

#### 定義: 微分係数
ある関数$f(x)$の$x=a$におけて

$$f'(a) = \lim_{ h \rightarrow 0} \frac{f(a+h) - f(a)}{ h }$$

が存在するとき、$f'(a)$を微分係数とする。

更に、導関数を定義します。


### 定義: 導関数

ある関数$f(x)$に対して、

$$f'(x)= \lim_{h \rightarrow 0} \frac{ f(x+h) - f(x)}{ h }$$

が存在するとき、$f'(x)$を導関数とする。

注意点として、微分係数はある一つの**値**であり、導関数は微分係数を求めるための**関数**です。
